document Data Storage (#378)
alpaka-group/alpaka3: docs/source/advanced/datastorage.rst

.. _advanced-data-storage:

Data Storage
============

The general structure of ``Data Storage`` is described in the section :ref:`Terms & Structure: Data Storage <basic-data-storage>`.
This page contains additional information about ``Data Storage``.

Using the Interface
-------------------

When a ``Data Storage`` interface concept is used in a function's argument list, it describes the **minimum** requirements that ``Data Storage`` objects must meet in order to be used to call the function.

.. literalinclude:: ../../snippets/dataStorage/datastorage_interface.cpp
  :language: cpp
  :start-after: BEGIN-DATASTORAGE-interface
  :end-before: END-DATASTORAGE-interface
  :dedent:

Read-only access via const annotation
-------------------------------------

> All objects that implements a ``Data Storage`` interface are const-correct.

Reviewer: ```suggestion
All objects that implement a ``Data Storage`` interface are const-correct.
```

If either the value type is ``const`` (e.g., ``alpaka::MdSpan<float const, TExtent, TPitch, TAlignment>``) or the ``Data Storage`` type itself is annotated with ``const`` (e.g., ``alpaka::MdSpan<float, TExtent, TPitch, TAlignment> const``), the values in the storage cannot be changed.

Special case: mutable alpaka::concepts::IDataSource
```````````````````````````````````````````````````

``alpaka::concepts::IDataSource`` only requires that data can be read from a ``Data Storage`` object, but not written to it, regardless of the ``const`` annotation.

> However, there are valid cases where a function argument is type-restricted with ``alpaka::concepts::IDataSource`` but requires that data can be written to it, which means that the actual ``Data Storage`` object must implement the ``alpaka::concepts::IMdSpan`` interface.

Reviewer: This is not really well explained yet I think.
```suggestion
However, there are valid cases where a function argument only requires ``alpaka::concepts::IDataSource`` in the general case but depending on other arguments can require mutability as well.
```

.. literalinclude:: ../../snippets/dataStorage/datastorage_writeable_datasource.cpp
  :language: cpp
  :start-after: BEGIN-DATASTORAGE-writeableDatasource
  :end-before: END-DATASTORAGE-writeableDatasource
  :dedent:

> `alpaka::onAcc::SimdAlgo.concurrent() <https://alpaka3.readthedocs.io/en/latest/doxygen/structalpaka_1_1onAcc_1_1SimdAlgo.html#a5face31ec9941f1eeb69fef9ea9fba01>`_ requires the ``IDataSource`` interface for the ``Data Storage`` object.

Reviewer: ```suggestion
For example, `alpaka::onAcc::SimdAlgo.concurrent() <https://alpaka3.readthedocs.io/en/latest/doxygen/structalpaka_1_1onAcc_1_1SimdAlgo.html#a5face31ec9941f1eeb69fef9ea9fba01>`_ requires the ``IDataSource`` interface for the ``Data Storage`` object.
```

The ``IDataSource`` interface only supports reading data. Depending on the user-defined functor, some of the ``Data Storage`` objects must be writable, so they must implement the ``IMdSpan`` interface.
However, the ``IMdSpan`` interface would prevent the user from using generators that only implement the ``IDataSource`` interface.
Therefore, the minimum requirement must be the non-constant ``IDataSource`` interface.

.. _memory-layout-of-multidimensional-data-storage:

Memory Layout of multidimensional Data Storage
----------------------------------------------

There are several functions and parameters for improving the memory layout of multidimensional ``Data Storage`` to enhance application performance.
Alpaka supports ``Pitches``, which optimize memory loads, and ``Alignment``, which is required for vector operations such as AVX on CPUs.
All alpaka functions automatically handle ``Pitches`` and ``Alignment`` during memory access.
However, it is sometimes necessary to process raw memory, for example, when a memory pointer is passed from alpaka to non-alpaka code.
The following section explains how alpaka implements ``Pitches`` and ``Alignment``.

Pitches
```````

> A ``Pitch`` consists of ``Extents`` (user data) plus some padding bytes to achieve a specific size in bytes for optimized memory loads.

Reviewer: I don't like the phrasing of this it seems misleading, but I can't really think of a better wording myself... The pitch doesn't really "consist" of extents.

Author: What do you think about?

The ``Pitch`` is the sum of user data (size of the ``Extent`` multiplied with the value type size) and some padding bytes to achieve a specific size in bytes for optimized memory loads.

Reviewer 2: `Pitches` are the number of bytes required to advance the pointer to the next element within each dimension of the memory.
The pitch for the columns (x-dimension) is the size of the value type.
For n-dimensional memory, the pitch to access the next memory row (y-dimension) is the number of bytes of a row to hold all elements in the x-dimension, plus optionally padding bytes required to align the address for the next row.

Reviewer: Yeah I think Rene's is good, it's explicit that the pitches don't actually contain the memory, they just tell you the shape of the buffer in physical memory

Author: My first sentence is wrong. I will overwork the first part of the documentation tomorrow.

Author: @AntonReinhard Can you please the check the `Pitch` documentation again.

For example, an Nvidia GPU can load 128 bytes with a single load command.
If we have a matrix with 32-bit integers (4 bytes) and 5 rows with 30 elements each, each row requires 120 bytes.
To ensure that each row can be loaded with a single load command and that no data from a second row is loaded, which would result in some rows requiring multiple load commands, 8 bytes of padding bytes are added.
The same principals are required for vectorization on CPU.

When the padding bytes are added, the ``Extents`` can no longer be used to calculate the memory location of a specific element.
The ``Extents`` assume that the memory is contiguous.
``Pitches`` solves this problem.
They stores the number of bytes required to jump to the next element in a dimension, including the padding bytes.
In the simplest case, when 1D ``Data Storage`` is used, the size of the value type in bytes is the same number which the ``Pitch`` stores.
For example, if the value type is float (32 bits), the ``Pitch`` is 4 bytes.
With 2D ``Data Storage``, the first number of a 2D ``Pitch`` stores the size of the row in bytes, i.e., ``number_of_elements * sizeof(value_type) + padding_bytes``.

> So if you have a memory pointer and add the ``Pitch[0]``, we jump one column further. ``Pitch[1]`` is again the size of the value type.

Reviewer 2: `Pitch[1] is again the size of the value type.` is wrong, it is the length of the row in bytes.

Reviewer 2: [1] is the size of the type, the sentense before is wrong and @SimeonEhrig fixed it.
